### PR TITLE
Add current_version

### LINKED ISSUE #107: Only allow HTTPS for API call?
``` php
        // lib/Brera/src/Api/Api.php:93
        if (empty($urlParts['scheme']) || $urlParts['scheme'] !== 'https') {
            // TODO comment
            #throw new InvalidHostException('Host should be called via HTTPS!');
        }
```

### LINKED ISSUE #328: Add data_versioning PUT/GET

### LINKED ISSUE #330: Update tests to PHPUnit 5.7

### LINKED ISSUE #331: Fix broken test \LizardsAndPumpkins\MagentoConnector\XmlBuilder\ListingXmlTest::testListingNodeContainsWebsiteAttribute

### LINKED ISSUE #332: Fix \LizardsAndPumpkins_MagentoConnector_Model_Resource_Catalog_Product_CollectionTest::testIncludesProductInCategoryUrlKeysAsNonCanonicalUrlKeys

### LINKED ISSUE #334: Fix Integration tests

### PR DESCRIPTION
**Pull request by Schrank**
FIXES #334
FIXES #332 
FIXES #331
FIXES #330
FIXES #328 
FIXES #107

### CODE REVIEW

File: tests/Unit/Suites/Api/ApiTest.php [hunk 1]
```diff
@@ -1,13 +1,47 @@
...
 {
     /**
+     * @var Request[]
+     */
+    private $requests;
+
+    /**
+     * @var string
+     */
+    private $host = 'https://api.lizardsandpumpkins.io/api';
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: .io domain?

File: tests/Unit/Suites/Api/ApiTest.php [hunk 2]
```diff
@@ -23,45 +57,65 @@ public function testWrongHost($url)
...
-            [null],
         ];
     }
 
     public function testValidHost()
     {
-        $host = 'https://api.lizardsAndPumpkins.io/api';
-        $api = new Api($host);
-        $this->assertInstanceOf(Api::class, $api);
+        $this->setUp();
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: The method `setUp` is called automatically before the test. If you call it manually from the test it will be executed twice. I think this most probably is not intended here, right?

File: tests/Unit/Suites/Model/Catalog/CategoryUrlKeyServiceTest.php [hunk 3]
```diff
@@ -46,7 +46,9 @@ class LizardsAndPumpkins_MagentoConnector_Model_Catalog_CategoryUrlKeyServiceTes
     private function createMockCategoryCollectionWithData(array $categoriesData)
     {
         $class = \LizardsAndPumpkins_MagentoConnector_Model_Resource_Catalog_Category_Collection::class;
-        $this->stubCategoryCollection = $this->getMock($class, [], [], '', false);
+        $this->stubCategoryCollection = $this->getMockBuilder($class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (Vinai) [inline tests/Unit/Suites/Model/Catalog/CategoryUrlKeyServiceTest.php]: `$this->createMock()` already disables the constructor.

File: tests/Unit/Suites/XmlBuilder/ListingXmlTest.php [hunk 4]
```diff
@@ -51,31 +56,40 @@ private function getListingAttributesAsArray($listingXmlString)
...
         $this->listingXml = new ListingXml($this->stubConfig);
 
-        $this->stubWebsite = $this->getMock(Mage_Core_Model_Website::class, [], [], '', false);
+        $this->stubWebsite = $this->getMockBuilder(Mage_Core_Model_Website::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->stubStore = $this->getMockBuilder(Mage_Core_Model_Store::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```
Comment (Vinai) [inline tests/Unit/Suites/XmlBuilder/ListingXmlTest.php]: `createMock()` is enough to disable the constructor.

File: tests/Unit/Suites/XmlBuilder/ListingXmlTest.php [hunk 5]
```diff
@@ -135,7 +149,8 @@ public function testListingNodeContainsStockAvailabilityCriteria()
     {
         $result = $this->listingXml->buildXml($this->stubCategory);
 
-        $expectedXml = <<<EOX
+        $expectedXml
+            = <<<EOX
```
Comment (Vinai) [inline tests/Unit/Suites/XmlBuilder/ListingXmlTest.php]: I think it looks nicer if the `=` is on the same line as the variable.

File: lib/LizardsAndPumpkins/src/Api/Api.php [hunk 6]
```diff
@@ -91,55 +51,112 @@ private function checkHost($url)
...
 
-    /**
-     * @param string $method
-     * @param string $url
-     * @param string[] $headers
-     * @param string $body
-     * @return Request
-     */
-    private function createHttpRequest($method, $url, $headers, $body)
+    public function setClient(Client $client)
```
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/Api.php]: Better to inject the client as a constructor dependency and not have a setter.

File: lib/LizardsAndPumpkins/src/Api/Api.php [hunk 7]
```diff
@@ -4,73 +4,33 @@
 
 use GuzzleHttp\Client;
```
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/Api.php]: i think it would be nice to wrap Guzzle into a small custom HttpApiClient class with an interface.
Th Api class then would know nothing about how the request is made, and creating the test double in the test would become trivial.

File: tests/Unit/Suites/Api/ApiTest.php [hunk 8]
```diff
@@ -23,45 +57,65 @@ public function testWrongHost($url)
...
-    public function testApiIsTriggered()
+    /**
+     * @see http://docs.guzzlephp.org/en/latest/testing.html
+     */
+    public function setUp()
     {
-        $this->markTestSkipped(
-            'No clue how to test this - but I\' pretty sure it is an implementation and not test problem'
-        );
+        $this->api = new Api($this->host);
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: Just inject a test double of the custom client interface here.
Then get rid of the HandlerStack and requests and so on. No real HttpClient needed at all during the test.

File: tests/Unit/Suites/XmlBuilder/ListingXmlTest.php [hunk 9]
```diff
@@ -51,31 +56,40 @@ private function getListingAttributesAsArray($listingXmlString)
...
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->stubStore = $this->getMockBuilder(Mage_Core_Model_Store::class)
+            ->disableOriginalConstructor()
+            ->getMock();
 
-        $stubStore = $this->getMock(Mage_Core_Model_Store::class, [], [], '', false);
-        $stubStore->method('getWebsite')->willReturn($this->stubWebsite);
+        $this->stubStore->method('getWebsite')->willReturn($this->stubWebsite);
```
Comment (Vinai) [inline tests/Unit/Suites/XmlBuilder/ListingXmlTest.php]: Is this still needed?

File: tests/Unit/Suites/XmlBuilder/ListingXmlTest.php [hunk 10]
```diff
@@ -107,7 +121,7 @@ public function testListingNodeContainsLocaleAttribute()
     public function testListingNodeContainsWebsiteAttribute()
     {
         $websiteCode = 'foo';
```
Comment (Vinai) [inline tests/Unit/Suites/XmlBuilder/ListingXmlTest.php]: Please rename `$websiteCode` to `$storeCode`.

File: lib/LizardsAndPumpkins/src/Api/Api.php [hunk 11]
```diff
@@ -4,73 +4,33 @@
...
 use GuzzleHttp\Psr7\Request;
+use function GuzzleHttp\json_decode;
```
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/Api.php]: PHPStorm was smarter than me :D Fixed.

File: lib/LizardsAndPumpkins/src/Api/Api.php [hunk 12]
```diff
@@ -4,73 +4,33 @@
...
-        $this->sendApiRequest($url, $headers, $body);
-    }
-
-    /**
-     * @param string $url
-     */
-    private function checkHost($url)
+    private function checkHost(string $url)
     {
         if (!is_string($url)) {
```
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/Api.php]: fixed

File: lib/LizardsAndPumpkins/src/Api/Api.php [hunk 13]
```diff
@@ -91,55 +51,112 @@ private function checkHost($url)
...
-    private function createHttpRequest($method, $url, $headers, $body)
+    public function setClient(Client $client)
     {
-        return new Request($method, $url, $headers, $body);
+        $this->client = $client;
+    }
+
+    public function triggerProductImport(string $filename)
+    {
+        $headers = ['Accept' => 'application/vnd.lizards-and-pumpkins.catalog_import.v1+json'];
```
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/Api.php]: yes

File: tests/Unit/Suites/Api/ApiTest.php (hunk 1 above)
Comment (Schrank) [inline tests/Unit/Suites/Api/ApiTest.php]: fixed

File: tests/Unit/Suites/Api/ApiTest.php (hunk 2 above)
Comment (Schrank) [inline tests/Unit/Suites/Api/ApiTest.php]: extracted method and left the junk. Thanks!

File: tests/Unit/Suites/XmlBuilder/ListingXmlTest.php (hunk 10 above)
Comment (Schrank) [inline tests/Unit/Suites/XmlBuilder/ListingXmlTest.php]: 👍

File: tests/Unit/Suites/XmlBuilder/ListingXmlTest.php (hunk 5 above)
Comment (Schrank) [inline tests/Unit/Suites/XmlBuilder/ListingXmlTest.php]: 👍

File: tests/Unit/Suites/XmlBuilder/ListingXmlTest.php (hunk 9 above)
Comment (Schrank) [inline tests/Unit/Suites/XmlBuilder/ListingXmlTest.php]: Yes, but the `$this->stubStore->method('getWebsite')->willReturn($this->stubWebsite);` 3 lines before is superflous

File: tests/Unit/Suites/Model/Catalog/CategoryUrlKeyServiceTest.php (hunk 3 above)
Comment (Schrank) [inline tests/Unit/Suites/Model/Catalog/CategoryUrlKeyServiceTest.php]: 👍

File: tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+        $this->request->method('getRouteName')->willReturn('adminhtml');
+        $this->request->method('getControllerName')->willReturn('lizardsandpumpkins_version');
+        $this->request->method('getRequestedRouteName')->willReturn('adminhtml');
+        $this->request->method('getRequestedControllerName')->willReturn('lizardsandpumpkins_version');
+        $this->request->method('isDispatched')->willReturn(true);
+
+        $this->response = $this->createMock(Zend_Controller_Response_Http::class);
+        $this->controller = new LizardsAndPumpkins_MagentoConnector_Adminhtml_LizardsAndPumpkins_VersionController(
+            $this->request, $this->response, [], $this->api
```
Comment (Vinai) [inline tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php]: Quoting PSR-2:

> Argument lists MAY be split across multiple lines, where each subsequent line is indented once. When doing so, the first item in the list MUST be on the next line, and there MUST be only one argument per line.

File: tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        );
+    }
+
+    public function testGetCurrentVersionAndSetOnBlock()
```
Comment (Vinai) [inline tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php]: I don't understand the test name.

File: tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+    {
+        $newVersion = uniqid('lap', true);
+        $this->api->expects($this->once())->method('setCurrentVersion')->with($this->equalTo($newVersion));
+
+        $this->dispatchUpdate($newVersion);
+    }
+
+    public function testSetVersionThrowsException()
+    {
+        $exceptionMessage = 'Request Failed was not catched!';
```
Comment (Vinai) [inline tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php]: To catch is an irregular verb (to catch, caught, caught): `"Request Failed" was not caught!`

File: tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+        $this->controller->dispatch('save');
+    }
+
+    private function dispatchUpdate(string $newVersion)
+    {
+        $this->request->method('getRequestedActionName')->willReturn('save');
+        $this->request->method('getActionName')->willReturn('save');
+
+        $this->request->method('getParam')->willReturnCallback(function ($param) use ($newVersion) {
+            if ($param === 'current_version') {
```
Comment (Vinai) [inline tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php]: Use yoda notation please with equality operators.

File: tests/Integration/Suites/Model/Resource/Catalog/Product/CollectionTest.php [hunk 18]
```diff
@@ -35,6 +37,8 @@ public function testItExtendsTheEavProductCollection()
 
     public function testIncludesProductInCategoryUrlKeysAsNonCanonicalUrlKeys()
     {
+        $this->markTestSkipped('FIX ME');
```
Comment (Vinai) [inline tests/Integration/Suites/Model/Resource/Catalog/Product/CollectionTest.php]: FIX ME

File: tests/Unit/Suites/Api/ApiTest.php [hunk 19]
```diff
@@ -1,67 +1,138 @@
 <?php
 
+declare(strict_types=1);
+
 namespace LizardsAndPumpkins\MagentoConnector\Api;
 
+use function GuzzleHttp\json_encode;
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: Why not use the native `\json_encode()` function?

File: tests/Unit/Suites/Api/ApiTest.php [hunk 20]
```diff
@@ -1,67 +1,138 @@
...
+        $previous = uniqid('previous-', true);
+        $responseBody = json_encode([
+            'data' => [
+                'current_version'  => $current,
+                'previous_version' => $previous,
+            ],
+        ]);
+
+        $headers = ['Accept' => 'application/vnd.lizards-and-pumpkins.current_version.v1+json'];
+        $url = $this->host . '/current_version/';
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: Better without a trailing `/`.

File: tests/Unit/Suites/Api/ApiTest.php [hunk 21]
```diff
@@ -1,67 +1,138 @@
...
+        $body = '';
+
+        $this->httpClient->expects($this->once())
+            ->method('getRequest')
+            ->with(
+                $this->equalTo($url),
+                $this->equalTo($body),
+                $this->equalTo($headers)
+            )->willReturn($responseBody);
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]:
```php
->with($url, $body, $headers)
```

File: tests/Unit/Suites/Api/ApiTest.php [hunk 22]
```diff
@@ -1,67 +1,138 @@
...
+        $headers = ['Accept' => 'application/vnd.lizards-and-pumpkins.current_version.v1+json'];
+        $url = $this->host . '/current_version/';
+        $body = ['current_version' => $newVersion];
+
+        $this->httpClient->expects($this->once())
+            ->method('putRequest')
+            ->with(
+                $this->equalTo($url),
+                $this->equalTo(json_encode($body)),
+                $this->equalTo($headers)
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: Here also the `$this->equalTo()` matchers aren't needed.

File: tests/Unit/Suites/Api/ApiTest.php [hunk 23]
```diff
@@ -1,67 +1,138 @@
...
         );
 
-        return;
+        $this->httpClient->expects($this->once())
+            ->method('putRequest')
+            ->with(
+                $this->equalTo($url),
+                $this->equalTo($body),
+                $this->equalTo($headers)
+            )->willReturn($responseBody);
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: No need for the `equalTo` matchers.

File: tests/Unit/Suites/Api/ApiTest.php [hunk 24]
```diff
@@ -1,67 +1,138 @@
...
+    {
+        $responseBody = '';
+
+        $headers = ['Accept' => 'application/vnd.lizards-and-pumpkins.catalog_import.v1+json'];
+        $url = $this->host . '/catalog_import/';
+        $body = json_encode(
+            [
+                'fileName' => 'catalog.xml',
+            ]
+        );
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: I'm sure this will fit nicely on one line.

File: tests/Unit/Suites/Api/ApiTest.php [hunk 25]
```diff
@@ -1,67 +1,138 @@
...
 
-        $file = 'catalog.xml';
+        $this->httpClient->expects($this->once())
+            ->method('putRequest')
+            ->with(
+                $this->equalTo($url),
+                $this->equalTo($body),
+                $this->equalTo($headers)
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: No need for the matchers if they are equalTo.

File: tests/Unit/Suites/Api/GuzzleAdapterTest.php [hunk 26]
```diff
@@ -0,0 +1,263 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LizardsAndPumpkins\MagentoConnector\Api;
+
+use GuzzleHttp\Client;
```
Comment (Vinai) [inline tests/Unit/Suites/Api/GuzzleAdapterTest.php]: Please alias `Guzzle\Client` as `GuzzleClient`.

File: tests/Unit/Suites/Api/GuzzleAdapterTest.php [hunk 27]
```diff
@@ -0,0 +1,263 @@
...
+
+        $this->client->putRequest($url, $body, $headers);
+    }
+
+    public function provideNon202StatusCodes(): array
+    {
+        return $this->provideStatusCodes([202]);
+    }
+
+    private function provideStatusCodes(array $except = []): array
```
Comment (Vinai) [inline tests/Unit/Suites/Api/GuzzleAdapterTest.php]: Rename to `private function getHttpStatusCodesExcept(int ...$except): array`

File: tests/Unit/Suites/XmlBuilder/ListingXmlTest.php [hunk 28]
```diff
@@ -51,31 +57,46 @@ private function getListingAttributesAsArray($listingXmlString)
...
+            }
+            if ($attribute === 'description') {
+                return 'Description with <strong>HTML</strong>';
+            }
+            if ($attribute === 'meta_description') {
+                return 'this is a meta description';
+            }
+            if ($attribute === 'meta_keywords') {
+                return 'meta keywords lap is cool';
+            }
```
Comment (Vinai) [inline tests/Unit/Suites/XmlBuilder/ListingXmlTest.php]: An alternative is `willReturnMap`:
```php
    $this->stubCategory->method('getData')->willReturnMap([
            ['meta_title', null, 'This would only work in a <CDATA> section'],
            ['description', null, 'Description with <strong>HTML</strong>'],
            ['meta_description', null, 'this is a meta description'],
            ['meta_keywords', null, 'meta keywords lap is cool'],
        ])
```
The second parameter `null` has to be present because of the optional second parameter to `getData()`.

File: magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Block/Adminhtml/Version/Form.php [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+        $fieldset = $form->addFieldset(
+            'fieldset',
+            ['legend' => Mage::helper('lizardsAndPumpkins_magentoconnector')->__('Current Version from API')]
+        );
+
+        $fieldset->addField('current_version', 'text', [
+            'label'    => Mage::helper('lizardsAndPumpkins_magentoconnector')->__('Current Version'),
+            'name'     => 'current_version',
+            'required' => true,
+            'value'    => $this->getVersion()['data']['current_version'],
```
Comment (Schrank) [inline magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Block/Adminhtml/Version/Form.php]: I thought the same but was not able to trigger that error, neither in test, nor on magento itself, but will change

File: tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php [hunk 30]
```diff
@@ -0,0 +1,179 @@
...
+    private $controller;
+
+    protected function setUp()
+    {
+        $adminSession = $this->getMockBuilder(Mage_Admin_Model_Session::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['isAllowed', 'getUser'])
+            ->getMock();
+        $adminSession->method('isAllowed')->willReturnCallback(function ($acl) {
+            if ($acl === 'system/index') {
```
Comment (Schrank) [inline tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php]: What is yoda notation? $var last, like 'string' === $var ?

File: tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php [hunk 31]
```diff
@@ -0,0 +1,179 @@
...
+        });
+        $adminSession->method('getUser')->willReturn($this->createMock(Mage_Admin_Model_User::class));
+
+        Mage::unregister('_singleton/admin/session');
+        Mage::register('_singleton/admin/session', $adminSession);
+
+        $this->adminhtmlSessionMock = $this->getMockBuilder(Mage_Adminhtml_Model_Session::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['addError'])
+            ->getMock();
```
Comment (Schrank) [inline tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php]: Unfortunately no - then I get weird errors I don't want to trace tbh

File: tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php (hunk 15 above)
Comment (Schrank) [inline tests/Integration/Suites/Controller/LizardsAndPumpkins/VersionControllerTest.php]: Is `testGetCurrentVersionFromApiAndSetOnBlock` better?

File: tests/Unit/Suites/Api/GuzzleAdapterTest.php (hunk 27 above)
Comment (Schrank) [inline tests/Unit/Suites/Api/GuzzleAdapterTest.php]: all my data providers start with provide - we should discuss this. I like it, when I immediately know, that a method is a data provider in a test

File: tests/Integration/Suites/Model/Resource/Catalog/Product/CollectionTest.php (hunk 18 above)
Comment (Schrank) [inline tests/Integration/Suites/Model/Resource/Catalog/Product/CollectionTest.php]: I'll try - please check extra careful, that I don't break anything.

Comment (Schrank) [inline tests/Integration/Suites/Model/Resource/Catalog/Product/CollectionTest.php]: I have no clue why this even worked in the first place. Down the stack trace it tries to `\LizardsAndPumpkins_MagentoConnector_Model_Resource_Catalog_Category_Collection::getRootCategoryIdPathForStore()` for the admin store. The store `0` lacks a root category. Someone who knows what is happening must have a look please. @tim-bezhashvyly @Vinai

**Comment by Schrank** [comment]
Pushed updated, thanks for the next review

File: lib/LizardsAndPumpkins/src/Api/GuzzleHttpApiClient.php [hunk 32]
```diff
@@ -27,11 +27,12 @@ public function putRequest(string $url, string $body, array $headers): string
         $method = 'PUT';
         $response = $this->sendRequest($url, $method, $body, $headers);
 
-        if ($response->getStatusCode() !== 202) {
+        if (((string)$response->getStatusCode())[0] !== '2') {
```
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/GuzzleHttpApiClient.php]: Um, I guess it would be okay, but this would match any number starting with a 2, and it's not really very nice, is it? It just doesn't seem like a proper robust check.
How about `($reponse->getStatusCode() >= 200 && $response->getStatusCode() < 300)`?

File: lib/LizardsAndPumpkins/src/Api/GuzzleHttpApiClient.php [hunk 33]
```diff
@@ -27,11 +27,12 @@ public function putRequest(string $url, string $body, array $headers): string
...
             throw new RequestFailedException(
-                sprintf('Status code %s does not match expected 202.', $response->getStatusCode())
+                sprintf('Status code %s does not match expected 200.', $response->getStatusCode())
```
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/GuzzleHttpApiClient.php]: The exception message should probably be something like `The HTTP response status code of the API is not within the expected 200-299 range, got %d`.

File: lib/LizardsAndPumpkins/src/Api/GuzzleHttpApiClient.php [hunk 34]
```diff
@@ -41,22 +42,22 @@ public function getRequest(string $url, string $body, array $headers): string
         $method = 'GET';
         $response = $this->sendRequest($url, $method, $body, $headers);
 
-        if ($response->getStatusCode() !== 200) {
+        if (((string)$response->getStatusCode())[0] !== '2') {
             throw new RequestFailedException(
-                sprintf('Status code %s does not match expected 200.', $response->getStatusCode())
+                sprintf('Status code %s does not match expected 2xx.', $response->getStatusCode())
```
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/GuzzleHttpApiClient.php]: See above where the code is pretty much duplicated.

File: lib/LizardsAndPumpkins/src/XmlBuilder/ProductBuilder.php [hunk 35]
```diff
@@ -67,11 +67,7 @@ private function buildProductXml(array $productXml)
                     continue;
                 }
 
-                $attributeNodeName = 'categories' == $attributeName
-                    ?
-                    'category'
-                    :
-                    $attributeName;
+                $attributeNodeName = 'categories' == $attributeName ? 'category' : $attributeName;
```
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/XmlBuilder/ProductBuilder.php]: Please use type safe comparison `===`, unless there is a good reason to use PHP type juggling.

File: magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Block/Adminhtml/Version/Form.php [hunk 36]
```diff
@@ -17,19 +17,19 @@ protected function _prepareForm()
...
         );
 
+        $versions = $this->getVersion()['data'];
         $fieldset->addField('current_version', 'text', [
-            'label'    => Mage::helper('lizardsAndPumpkins_magentoconnector')->__('Current Version'),
+            'label'    => Mage::helper('lizardsAndPumpkins_magentoconnector')->__('Current Data Version from API'),
             'name'     => 'current_version',
             'required' => true,
-            'value'    => $this->getVersion()['data']['current_version'],
+            'value'    => isset($versions['current_version']) ? $versions['current_version'] : '',
```
Comment (Vinai) [inline magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Block/Adminhtml/Version/Form.php]: The PHP7 uncertainty operator would work well here: `=> $versions['current_version'] ?? ''`;

File: magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Block/Adminhtml/Version/Form.php [hunk 37]
```diff
@@ -17,19 +17,19 @@ protected function _prepareForm()
...
         ]);
 
         $fieldset->addField('previous_version', 'text', [
-            'label'    => Mage::helper('lizardsAndPumpkins_magentoconnector')->__('Previous Version'),
+            'label'    => Mage::helper('lizardsAndPumpkins_magentoconnector')->__('Previous Data Version'),
             'name'     => 'previous_version',
-            'required' => true,
             'readonly' => true,
-            'value'    => $this->getVersion()['data']['previous_version'],
+            'value'    => isset($versions['previous_version']) ? $versions['previous_version'] : '',
```
Comment (Vinai) [inline magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Block/Adminhtml/Version/Form.php]: See above, `??` is your friend :)

File: magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/controllers/Adminhtml/LizardsAndPumpkins/VersionController.php [hunk 38]
```diff
@@ -49,8 +49,8 @@ public function saveAction()
         $session = Mage::getSingleton('adminhtml/session');
 
         $version = $this->getRequest()->getParam('current_version');
-        if (!is_string($version) || $version === '') {
-            $session->addError('Current version must not be empty!');
+        if (!is_string($version) || '' === $version) {
```
Comment (Vinai) [inline magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/controllers/Adminhtml/LizardsAndPumpkins/VersionController.php]: Please `trim($version)`.

File: tests/Unit/Suites/Api/ApiTest.php [hunk 39]
```diff
@@ -44,15 +42,15 @@ public function testApiGetCurrentVersion()
...
 
         $this->httpClient->expects($this->once())
             ->method('getRequest')
             ->with(
-                $this->equalTo($url),
-                $this->equalTo($body),
-                $this->equalTo($headers)
+                $url,
+                $body,
+                $headers
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: Doesn't it fit in one line, e.g. `->with($url, $body, $headers)`

File: tests/Unit/Suites/Api/ApiTest.php [hunk 40]
```diff
@@ -65,15 +63,15 @@ public function testApiSetCurrentVersion()
...
 
         $this->httpClient->expects($this->once())
             ->method('putRequest')
             ->with(
-                $this->equalTo($url),
-                $this->equalTo(json_encode($body)),
-                $this->equalTo($headers)
+                $url,
+                json_encode($body),
+                $headers
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: See above, there isn't a real reason to have the 3 arguments on separate lines.

File: tests/Unit/Suites/Api/ApiTest.php [hunk 41]
```diff
@@ -102,9 +100,9 @@ public function testTriggerCmsBlockUpdate()
         $this->httpClient->expects($this->once())
             ->method('putRequest')
             ->with(
-                $this->equalTo($url),
-                $this->equalTo($body),
-                $this->equalTo($headers)
+                $url,
+                $body,
+                $headers
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: See above, `Ctl-Shift-J Ctl-Shift-J Ctl-Shift-J Ctl-Shift-J`

File: tests/Unit/Suites/Api/ApiTest.php [hunk 42]
```diff
@@ -116,23 +114,17 @@ public function testTriggerProductImport()
...
 
         $this->httpClient->expects($this->once())
             ->method('putRequest')
             ->with(
-                $this->equalTo($url),
-                $this->equalTo($body),
-                $this->equalTo($headers)
+                $url,
+                $body,
+                $headers
```
Comment (Vinai) [inline tests/Unit/Suites/Api/ApiTest.php]: Time to save some bytes.

File: tests/Unit/Suites/Api/GuzzleHttpApiClientTest.php [hunk 43]
```diff
@@ -124,9 +119,9 @@ public function testThrowsExceptionOnNon202PutResponse(int $statusCode)
         $this->client->putRequest($url, $body, $headers);
     }
 
-    public function provideNon202StatusCodes(): array
+    public function provideNon2xxStatusCodes(): array
     {
-        return $this->provideStatusCodes([202]);
+        return $this->provideStatusCodes([200, 201, 202, 203, 204, 205, 206, 207, 208, 226]);
```
Comment (Vinai) [inline tests/Unit/Suites/Api/GuzzleHttpApiClientTest.php]: I still think it would be better to make `provideStatusCodes()` to use variadic arguments...

File: lib/LizardsAndPumpkins/src/XmlBuilder/ProductBuilder.php (hunk 35 above)
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/XmlBuilder/ProductBuilder.php]: I was tempted too, but the file is not part of the PR... Fixed it :)

File: tests/Unit/Suites/Api/GuzzleHttpApiClientTest.php (hunk 43 above)
Comment (Vinai) [inline tests/Unit/Suites/Api/GuzzleHttpApiClientTest.php]: Example:
```php
public function provideNon2xxStatusCodes(): array
{
    return $this->provideStatusCodesExcept(200, 201, 202, 203, 204, 205, 206, 207, 208, 226); // <-- no array argument, just INTs
}

private function provideStatusCodesExcept(int ...$except): array // <-- $except now is array of INT
{
```

Comment (Vinai) [inline tests/Unit/Suites/Api/GuzzleHttpApiClientTest.php]: Renaming the method to `provideStatusCodesExcept` also fixes the problem that it does the opposite of providing the specified status codes .

**Comment by Schrank** [comment]
Ok, another push :)
